File: buildsys.go
package main

import "path/filepath"

type buildsysKind int

const (
	buildNone buildsysKind = iota
	buildCMake
	buildMake
	buildAutotools
)

// buildsysDetect detects the build system in dir by checking for marker
// files, in priority order: CMakeLists.txt -> cmake; an executable
// "configure" script -> autotools; a Makefile -> make.
func buildsysDetect(dir string) buildsysKind {
	if pathExists(filepath.Join(dir, "CMakeLists.txt")) {
		return buildCMake
	}
	if pathIsExecutableFile(filepath.Join(dir, "configure")) {
		return buildAutotools
	}
	if pathExists(filepath.Join(dir, "Makefile")) || pathExists(filepath.Join(dir, "makefile")) {
		return buildMake
	}
	return buildNone
}

func buildsysName(kind buildsysKind) string {
	switch kind {
	case buildCMake:
		return "CMake"
	case buildMake:
		return "Make"
	case buildAutotools:
		return "Autotools"
	default:
		return "none"
	}
}

func buildCmakeImpl(dir string, extraFlags []string) int {
	builddir := filepath.Join(dir, "build")

	argv := append([]string{"cmake", "-S", dir, "-B", builddir}, extraFlags...)
	// Configure/compile output is captured behind a spinner instead of
	// flooding the terminal -- only shown in full on failure, when it's
	// the only useful diagnostic. The install step stays plain
	// passthrough: it needs sudo, and sudo's password prompt goes
	// straight to the controlling terminal, which would visually collide
	// with an animated spinner redrawing the same line.
	rc := runCommandSpinner("", argv, "configuring with CMake...", false)
	if rc != 0 {
		printErr("cmake configure failed")
		return rc
	}
	printOK("configured")

	rc = runCommandSpinner("", []string{"cmake", "--build", builddir}, "building", true)
	if rc != 0 {
		printErr("cmake build failed")
		return rc
	}
	printOK("compiled")

	printInfo("installing (sudo cmake --install)...")
	rc = runCommand("", []string{"sudo", "cmake", "--install", builddir})
	if rc != 0 {
		printErr("cmake --install failed")
	}
	return rc
}

func buildMakeImpl(dir string) int {
	rc := runCommandSpinner(dir, []string{"make"}, "building with make...", false)
	if rc != 0 {
		printErr("make failed")
		return rc
	}
	printOK("compiled")

	printInfo("installing (sudo make install)...")
	rc = runCommand(dir, []string{"sudo", "make", "install"})
	if rc != 0 {
		// Don't guess at a specific cause (missing install target, a file
		// the install step references being absent, a permissions
		// issue, ...) -- make already printed the real reason above;
		// repeating a wrong guess is worse than saying nothing.
		printErr("'make install' failed -- see output above for the reason.")
	}
	return rc
}

func buildAutotoolsImpl(dir string) int {
	rc := runCommandSpinner(dir, []string{"./configure"}, "running ./configure...", false)
	if rc != 0 {
		printErr("./configure failed")
		return rc
	}
	printOK("configured")
	return buildMakeImpl(dir)
}

// buildsysBuildAndInstall builds and installs the project in dir using
// the given build system. extraCmakeFlags are appended to the cmake
// configure step only; other build systems ignore them for now. Install
// steps run under sudo, since they write outside the user's home
// directory.
func buildsysBuildAndInstall(kind buildsysKind, dir string, extraCmakeFlags []string) int {
	switch kind {
	case buildCMake:
		return buildCmakeImpl(dir, extraCmakeFlags)
	case buildMake:
		return buildMakeImpl(dir)
	case buildAutotools:
		return buildAutotoolsImpl(dir)
	default:
		printErr("internal error: no build system to run")
		return -1
	}
}

File: src/buildsys.c
#include "buildsys.h"

#include <stdio.h>
#include <stdlib.h>

#include "proc.h"
#include "util.h"

buildsys_kind_t buildsys_detect(const char *dir) {
    char *path;

    path = path_join(dir, "CMakeLists.txt");
    int has_cmake = path_exists(path);
    free(path);
    if (has_cmake) return BUILD_CMAKE;

    path = path_join(dir, "configure");
    int has_configure = path_is_executable_file(path);
    free(path);
    if (has_configure) return BUILD_AUTOTOOLS;

    path = path_join(dir, "Makefile");
    int has_makefile = path_exists(path);
    free(path);
    if (has_makefile) return BUILD_MAKE;

    path = path_join(dir, "makefile");
    has_makefile = path_exists(path);
    free(path);
    if (has_makefile) return BUILD_MAKE;

    return BUILD_NONE;
}

const char *buildsys_name(buildsys_kind_t kind) {
    switch (kind) {
        case BUILD_CMAKE: return "CMake";
        case BUILD_MAKE: return "Make";
        case BUILD_AUTOTOOLS: return "Autotools";
        default: return "none";
    }
}

static int build_cmake(const char *dir, char **extra_flags, size_t n_extra) {
    char *builddir = path_join(dir, "build");

    size_t argc = 5 + n_extra + 1;
    char **argv = xmalloc(argc * sizeof(char *));
    size_t i = 0;
    argv[i++] = "cmake";
    argv[i++] = "-S";
    argv[i++] = (char *)dir;
    argv[i++] = "-B";
    argv[i++] = builddir;
    for (size_t j = 0; j < n_extra; j++) argv[i++] = extra_flags[j];
    argv[i++] = NULL;

    printf("goget: configuring with CMake...\n");
    int rc = run_command(argv);
    free(argv);
    if (rc != 0) {
        fprintf(stderr, "goget: cmake configure failed\n");
        free(builddir);
        return rc;
    }

    printf("goget: building...\n");
    char *const build_argv[] = {"cmake", "--build", builddir, NULL};
    rc = run_command(build_argv);
    if (rc != 0) {
        fprintf(stderr, "goget: cmake build failed\n");
        free(builddir);
        return rc;
    }

    printf("goget: installing (sudo cmake --install)...\n");
    char *const install_argv[] = {"sudo", "cmake", "--install", builddir, NULL};
    rc = run_command(install_argv);
    free(builddir);
    if (rc != 0) {
        fprintf(stderr, "goget: cmake --install failed\n");
    }
    return rc;
}

static int build_make(const char *dir) {
    printf("goget: building with make...\n");
    char *const make_argv[] = {"make", NULL};
    int rc = run_command_in(dir, make_argv);
    if (rc != 0) {
        fprintf(stderr, "goget: make failed\n");
        return rc;
    }

    printf("goget: installing (sudo make install)...\n");
    char *const install_argv[] = {"sudo", "make", "install", NULL};
    rc = run_command_in(dir, install_argv);
    if (rc != 0) {
        fprintf(stderr,
                "goget: 'make install' failed -- this project may not "
                "define an install target.\n");
    }
    return rc;
}

static int build_autotools(const char *dir) {
    printf("goget: running ./configure...\n");
    char *const configure_argv[] = {"./configure", NULL};
    int rc = run_command_in(dir, configure_argv);
    if (rc != 0) {
        fprintf(stderr, "goget: ./configure failed\n");
        return rc;
    }

    return build_make(dir);
}

int buildsys_build_and_install(buildsys_kind_t kind, const char *dir,
                                 char **extra_cmake_flags, size_t n_extra_flags) {
    switch (kind) {
        case BUILD_CMAKE:
            return build_cmake(dir, extra_cmake_flags, n_extra_flags);
        case BUILD_MAKE:
            return build_make(dir);
        case BUILD_AUTOTOOLS:
            return build_autotools(dir);
        default:
            fprintf(stderr, "goget: internal error: no build system to run\n");
            return -1;
    }
}
